### PR TITLE
Add Picture-in-picture

### LINKED ISSUE #174: [Feature]: Pip (Picture In Picture)
### What feature do you want to add?

A new button to the video player for Picture In Picture mode. 
image

https://www.w3.org/TR/picture-in-picture/
https://css-tricks.com/an-introduction-to-the-picture-in-picture-web-api/
https://googlechrome.github.io/samples/picture-in-picture/

### Why do you want to have this feature?

To allow watching of movies while on other tabs

### Anything other details to share?

_No response_

### PR DESCRIPTION
**Pull request by zisra**
Should close #174
**Q:** Would you prefer to hide or blur the PIP button if it is not available?

### CODE REVIEW

**Comment by mrjvs** [comment]
Hiding the button is good.

It also needs support for safari: https://developer.apple.com/documentation/webkitjs/adding_picture_in_picture_to_your_safari_media_controls#2940030

**Comment by zisra** [comment]
@mrjvs If the button is hidden if there is no support, would it make sense to make this consistent with the download button?

**Comment by mrjvs** [comment]
With download, its the content that isn't supported, not the browser.

We hide functionality quite a lot if browser doesnt support: volume, fullscreen, chromecast, airplay.

**Comment by JamesHawkinss** [comment]
This doesn't seem to be building

File: public/config.js [hunk 1]
```diff
@@ -1,7 +1,6 @@
 window.__CONFIG__ = {
   // url must NOT end with a slash
-  VITE_CORS_PROXY_URL: "",
-
+  VITE_CORS_PROXY_URL: "https://rough.isra.workers.dev",
```
Comment (JamesHawkinss) [inline public/config.js:3]: Do not commit your proxy URL

**Review by JamesHawkinss — CHANGES_REQUESTED** [review]
Still failing to build.

`12:23:20 PM: src/video/state/providers/castingStateProvider.ts(152,19): error TS2339: Property 'togglePictureInPicture' does not exist on type 'RemotePlayerController'.`

File: public/config.js (hunk 1 above)
Comment (zisra) [reply public/config.js:3]: I will fix it sorry

**Comment by zisra** [comment]
This should work

File: src/video/components/actions/PictureInPictureAction.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  }, [controls]);
+
+  if (!canPictureInPicture() && !canWebkitPictureInPicture()) return null;
+
+  return (
+    <VideoPlayerIconButton
+      className={props.className}
+      icon={Icons.PICTURE_IN_PICTURE}
+      onClick={handleClick}
+      disabled={false}
```
Comment (mrjvs) [inline src/video/components/actions/PictureInPictureAction.tsx:35]: remove disabled

File: src/video/state/providers/castingStateProvider.ts [hunk 3]
```diff
@@ -83,6 +83,9 @@ export function createCastingStateProvider(
       state.pausedWhenSeeking = state.mediaPlaying.isPaused;
       this.pause();
     },
+    togglePictureInPicture() {
+      updateSource(descriptor, state);
```
Comment (mrjvs) [inline src/video/state/providers/castingStateProvider.ts:87]: no need to update source if nothing happened

File: src/video/state/types.ts [hunk 4]
```diff
@@ -30,6 +30,7 @@ export type VideoPlayerState = {
     isFocused: boolean; // is the video player the users focus? (shortcuts only works when its focused)
     leftControlHovering: boolean; // is the cursor hovered over the left side of player controls
     popoutBounds: null | DOMRect; // bounding box of current popout
+    isPictureInPicture: boolean; // is picture in picture active
```
Comment (mrjvs) [inline src/video/state/types.ts:33]: the isPictureInPicture is not used anywhere, just remove it

**Review by mrjvs — CHANGES_REQUESTED** [review]
just a few changes needed.
I've tested everything except on safari.

**Review by JamesHawkinss — APPROVED** [review]

**Review by mrjvs — APPROVED** [review]

**Pull request merged**